docs/state_encryption.md

```diff
@@ -0,0 +1,333 @@
+# State encryption
+
+This document details our intended implementation of the state and plan encryption feature.
+
+## Notes
+
+This document mentions `HCL` as a short-form for OpenTofu code. Unless otherwise specified, everything written also applies to the [JSON-equivalent of the HCL code](https://opentofu.org/docs/language/syntax/json/).
+
+## Goals
+
+The goal of this feature is to allow OpenTofu users to fully encrypt state files when they are stored on the local disk or transferred to a remote backend. The feature should also allow reading from an encrypted remote backend using the `terraform_remote_state` data source. The encrypted version should still be a valid JSON file, but not necessarily a valid state file.
+
+Furthermore, this feature should allow users to encrypt plan files when they are stored. However, plan files are not JSON, they are ZIP files. If feasible, this feature should produce a structurally correct encrypted ZIP file.
+
+For the encryption key, users should be able to specify a key directly, use a remote key provider (such as AWS KMS, etc.), or create derivative keys (such as pbkdf2) from another key source. The primary encryption method should be AES-GCM, but the implementation should be open to different encryption methods. The user should also have the ability to decrypt a state or plan file with one (older) key and then re-encrypt data with a newer key. Multiple fallbacks should be avoided in the implementation.
```

Comment:
Could you expand on the use-case for doing derivative keys from another key source? I.e. with pbkdf2 the point is to use a human-readable password that is then expanded to a full key. I don't think there's any point in doing this for automated systems (which should just provide an AES key)?

Reply:
In the case of pbkdf2, using a non-human-readable password has always been best practice as far as I'm aware, to avoid brute-force attacks.

In this case, I think it could also be helpful if we have a 1password or similar key provider to pull the password from there.

Reply:
If you're already doing a non-human-readable password (so don't use the approach of https://xkcd.com/936/) then you can just as well go with a raw AES key.